File: greedy/1793-maximum-score-of-a-good-subarray.go
package greedy

// https://leetcode.com/problems/maximum-score-of-a-good-subarray/

/*
Lets say that nums[k]=7 and consider following casses.
You want to increase one end of your interval by one. Which side you should choose?
Lets increse the interval end points one at a time:
[.............., 3, 7, 6, ..........................] 3 and 6 are candidates. Choose 6
[............... 3, 7, 6, 4, .......................] 3 and 4 are candidates. Choose 4
[.............., 3, 7, 6, 4, 7,.....................]
[.............., 3, 7, 6, 4, 7, 2...................]
[.............4, 3, 7, 6, 4, 7, 2...................]
[..........2, 4, 3, 7, 6, 4, 7, 2...................]
[.......3, 2, 4, 3, 7, 6, 4, 7, 2, 1................]
Note: Edge cases are when leftEnd = 0 and rightEnd = len(nums) - 1
*/

func maximumScore(nums []int, k int) int {
	var n = len(nums)
	var score = nums[k]
	var minimum = nums[k]
	var left, right = k, k
	var lnum, rnum int
	for left > 0 || right+1 < n {
		lnum, rnum = 0, 0
		if left > 0 {
			lnum = nums[left-1]
		}
		if right+1 < n {
			rnum = nums[right+1]
		}
		if lnum > rnum {
			left--
			minimum = min(minimum, lnum)
		} else {
			right++
			minimum = min(minimum, rnum)
		}
		score = max(score, minimum*(right-left+1))
	}
	return score
}

File: greedy/1793-maximum-score-of-a-good-subarray.py
from typing import List
import unittest

# https://leetcode.com/problems/maximum-score-of-a-good-subarray/

# python3 -m unittest greedy/1793-maximum-score-of-a-good-subarray.py


class Solution(unittest.TestCase):
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        minimum = nums[k]
        score = nums[k]
        left = right = k
        while left > 0 or right + 1 < n:
            lnum = nums[left - 1] if left > 0 else 0
            rnum = nums[right + 1] if right + 1 < n else 0
            if lnum > rnum:
                left -= 1
                minimum = min(minimum, lnum)
            else:
                right += 1
                minimum = min(minimum, rnum)
            score = max(score, minimum * (right - left + 1))
        return score

    def test(self):
        for nums, k, expected in [
            ([1, 4, 3, 7, 4, 5], 3, 15),
            ([5, 5, 4, 5, 4, 1, 1, 1], 0, 20),
            ([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 2], 28, 30),
        ]:
            output = self.maximumScore(nums, k)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
